src/full/Agda/TypeChecking/Pretty/Warning.hs

```diff
@@ -690,15 +690,15 @@ tcWarningsToError mws = case (unsolvedHoles, otherWarnings) of
 -- | Depending which flags are set, one may happily ignore some
 -- warnings.
 
-applyFlagsToTCWarningsPreserving :: HasOptions m => Set WarningName -> [TCWarning] -> m [TCWarning]
+applyFlagsToTCWarningsPreserving :: HasOptions m => Set WarningName -> Set TCWarning -> m (Set TCWarning)
 applyFlagsToTCWarningsPreserving additionalKeptWarnings ws = do
   -- For some reason some SafeFlagPragma seem to be created multiple times.
   -- This is a way to collect all of them and remove duplicates.
-  let pragmas w = case tcWarning w of { SafeFlagPragma ps -> ([w], ps); _ -> ([], []) }
-  let sfp = case fmap List.nub (foldMap pragmas ws) of
-              (TCWarning loc r w p b:_, sfp) ->
-                 [TCWarning loc r (SafeFlagPragma sfp) p b]
-              _                        -> []
+  let pragmas w = case tcWarning w of { SafeFlagPragma ps -> ([w], ps); _ -> ([], empty) }
+  let sfp = case foldMap pragmas ws of
+              (TCWarning loc r w doc str b : _, sfp) ->
+                 singleton $ TCWarning loc r (SafeFlagPragma sfp) doc str b
+              _ -> empty
```

Comment:
Can we get rid entirely of this special logic for `SafeFlagPragma`?  One of these possibilities:
1. Make `SafeFlagPragma` take a single pragma name rather than a Set.
2. Store the `SafeFlagPragma` warning separately so that in the state these warnings are always properly aggregated.

Seems like 1. would be simplest.
Ad 2: We might want to aggregate other warnings as well (at least those that are deserialized), so we should have a uniform mechanism for that rather than this special case for `SafeFlagPragma`.